spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -777,17 +785,25 @@ public boolean canAdd(Message<?> message) {
 			Integer messageSequenceNumber = messageHeaderAccessor.getSequenceNumber();
 			if (messageSequenceNumber != null && messageSequenceNumber > 0) {
 				Integer messageSequenceSize = messageHeaderAccessor.getSequenceSize();
-				return messageSequenceSize.equals(this.getSequenceSize())
-						&& !this.containsSequenceNumber(this.getMessages(), messageSequenceNumber);
+				return messageSequenceSize.equals(getSequenceSize())
+						&& !(this.sourceGroup != null ? this.sourceGroup.containsSequence(messageSequenceNumber)
+								: containsSequenceNumber(this.getMessages(), messageSequenceNumber));
 			}
 			return true;
 		}
 
 		private boolean containsSequenceNumber(Collection<Message<?>> messages, Integer messageSequenceNumber) {
 			for (Message<?> member : messages) {
-				Integer memberSequenceNumber = new IntegrationMessageHeaderAccessor(member).getSequenceNumber();
-				if (messageSequenceNumber.equals(memberSequenceNumber)) {
-					return true;
+				Object sequenceNumber = member.getHeaders().get(
+						IntegrationMessageHeaderAccessor.SEQUENCE_NUMBER);
```

Comment:
Yes, I thought not one time that use extra object to just get friendly accessor to the value isn't good idea at all...
Let's don't use it in the `canAdd()` as well !

Anyway why `member.getHeaders().get(IntegrationMessageHeaderAccessor.SEQUENCE_NUMBER, Integer.class)` doesn't work for you here?
Looks like pretty parallel with what you do manually with this `if..else`